File: hash/subarray_0_sum.cpp
//{ Driver Code Starts
// A C++ program to find if there is a zero sum
// subarray
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
  // Complete this function
  // Function to check whether there is a subarray present with 0-sum or not.
  bool subArrayExists(int arr[], int n) {
    // Your code here
    if (n == 0)
      return false;
    unordered_set<int> s;
    int t = 0;
    s.insert(0);
    for (int i = 0; i < n; ++i) {
      t += arr[i];
      if (s.find(t) != s.end()) {
        return true;
      }
      s.insert(t);
    }

    return false;
  }
};

//{ Driver Code Starts.
// Driver code
int main() {
  int t;
  cin >> t;
  while (t--) {
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++)
      cin >> arr[i];
    Solution obj;
    if (obj.subArrayExists(arr, n))
      cout << "Yes\n";
    else
      cout << "No\n";
  }
  return 0;
}
// } Driver Code Ends

File: hash/subarray_0_sum.py
# User function Template for python3
from collections import defaultdict


class Solution:
    # Function to check whether there is a subarray present with 0-sum or not.
    def subArrayExists(self, arr, n):
        ##Your code here
        # Return true or false

        d = defaultdict(int)
        d[0] = 1
        curr = 0
        for i in range(n):
            curr += arr[i]
            if curr in d:
                return 1
            d[curr] = 1

        return 0


# {
# Driver Code Starts
# Initial Template for Python 3


def main():
    T = int(input())
    while T > 0:
        n = int(input())
        arr = [int(x) for x in input().strip().split()]
        if Solution().subArrayExists(arr, n):
            print("Yes")
        else:
            print("No")

        T -= 1


if __name__ == "__main__":
    main()
# } Driver Code Ends
